File: androiddemo/app/src/main/java/com/cdc/androiddemo/MainActivity.java
package com.cdc.androiddemo;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;

import com.cdc.mlog.MLog;

public class MainActivity extends AppCompatActivity {

    RecyclerView rv;
    Button nextBtn, prevBtn;
    Paginator p = new Paginator();
    private int totalPages = Paginator.TOTAL_NUM_ITEMS / Paginator.ITEMS_PER_PAGE;
    private int currentPage = 0;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        setContentView(R.layout.activity_main);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        //RFERENCE VIEWS
        rv = (RecyclerView) findViewById(R.id.rv);
        nextBtn = (Button) findViewById(R.id.nextBtn);
        prevBtn = (Button) findViewById(R.id.prevBtn);
        prevBtn.setEnabled(false);

        //RECYCLER PROPERTIES
        rv.setLayoutManager(new LinearLayoutManager(this));

        //ADAPTER
        rv.setAdapter(new MyAdapter(MainActivity.this, p.generatePage(currentPage)));

        //NAVIGATE
        nextBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {

                currentPage += 1;
                // enableDisableButtons();
                rv.setAdapter(new MyAdapter(MainActivity.this, p.generatePage(currentPage)));
                toggleButtons();

            }
        });
        prevBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                currentPage -= 1;

                rv.setAdapter(new MyAdapter(MainActivity.this, p.generatePage(currentPage)));

                toggleButtons();
            }
        });
    }

    private void toggleButtons() {
        if (currentPage == totalPages) {
            nextBtn.setEnabled(false);
            prevBtn.setEnabled(true);
        } else if (currentPage == 0) {
            prevBtn.setEnabled(false);
            nextBtn.setEnabled(true);
        } else if (currentPage >= 1 && currentPage <= totalPages) {
            nextBtn.setEnabled(true);
            prevBtn.setEnabled(true);
        }
    }
}

File: Okhttp/app/src/main/java/com/cdc/okhttp3/MainActivity.kt
package com.cdc.okhttp3
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.view.View
import kotlinx.android.synthetic.main.activity_main.*
import android.widget.Toast
import okhttp3.*
import okhttp3.logging.HttpLoggingInterceptor
import java.io.IOException
import java.util.concurrent.TimeUnit

class MainActivity : AppCompatActivity(), View.OnClickListener {
    val MEDIA_TYPE_MARKDOWN = MediaType.parse("text/x-markdown; charset=utf-8")
    private val MEDIA_TYPE_PNG = MediaType.parse("image/png")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)


        bt_send.setOnClickListener(this)
        bt_postsend.setOnClickListener(this)
        bt_sendfile.setOnClickListener(this)
        bt_downfile.setOnClickListener(this)
    }

    override fun onClick(v: View?) {
        when (v?.id) {
            R.id.bt_send -> {
                getAsynHttp()
            }
            R.id.bt_postsend -> {

            }
            R.id.bt_sendfile -> {

            }
            R.id.bt_downfile -> {

            }
        }
    }

    /**
     * get异步请求
     */
    private fun getAsynHttp() {
        //Log Info,Warn,Error可以看到日志;Log Verbose Debug看不到日志
        val logging = HttpLoggingInterceptor(HttpLoggingInterceptor.Logger {
            Log.i("ppp",it)
        })
        logging.redactHeader("Server")
        //val logging = HttpLoggingInterceptor()  看不到日志
        logging.level = HttpLoggingInterceptor.Level.BODY
        var mOkHttpClient=OkHttpClient.Builder()
                .addInterceptor(logging)
                .connectTimeout(15, TimeUnit.SECONDS)
                .writeTimeout(20, TimeUnit.SECONDS)
                .readTimeout(20, TimeUnit.SECONDS)
                .build()
        var requestBuilder = Request.Builder().url("https://api.github.com/users/cdcdec")
        var request = requestBuilder.build()
        var mcall = mOkHttpClient?.newCall(request)
        mcall?.enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                runOnUiThread { Toast.makeText(applicationContext, "请求失败:${call.request().url()}", Toast.LENGTH_SHORT).show() }
            }
            override fun onResponse(call: Call, response: Response) {
                var str = ""
                if (null != response.cacheResponse()) {
                    str = response.cacheResponse().toString()+"==cache"
                   // Log.i("wangshu", "cache---$str")
                } else {
                    response.body()?.string()
                    str = response.networkResponse().toString()+"==network"
                    //Log.i("wangshu", "network---$str")
                }
                runOnUiThread { Toast.makeText(applicationContext, "请求成功:$str", Toast.LENGTH_SHORT).show() }
            }
        })
    }
}
